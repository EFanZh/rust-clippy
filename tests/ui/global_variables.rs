#![warn(clippy::global_variables)]

use std::sync::Mutex;
use std::sync::atomic::AtomicU32;

macro_rules! define_global_variable_with_macro {
    () => {
        static GLOBAL_VARIABLE_0: AtomicU32 = AtomicU32::new(2);
        //~^ global_variables

        static GLOBAL_VARIABLE_1: Mutex<u32> = Mutex::new(3);
        //~^ global_variables
    };
}

fn main() {
    define_global_variable_with_macro!();

    static GLOBAL_VARIABLE_2: AtomicU32 = AtomicU32::new(2);
    //~^ global_variables

    static GLOBAL_VARIABLE_3: Mutex<u32> = Mutex::new(3);
    //~^ global_variables

    static NOT_GLOBAL_VARIABLE_0: u32 = 1;
    static NOT_GLOBAL_VARIABLE_1: Vec<AtomicU32> = Vec::new();
    static NOT_GLOBAL_VARIABLE_2: Vec<Mutex<u32>> = Vec::new();
}
