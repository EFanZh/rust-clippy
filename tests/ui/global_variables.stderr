error: global variable should not be used
  --> tests/ui/global_variables.rs:8:9
   |
LL |         static GLOBAL_VARIABLE_0: AtomicU32 = AtomicU32::new(2);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
LL |     define_global_variable_with_macro!();
   |     ------------------------------------ in this macro invocation
   |
   = note: `-D clippy::global-variables` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::global_variables)]`
   = note: this error originates in the macro `define_global_variable_with_macro` (in Nightly builds, run with -Z macro-backtrace for more info)

error: global variable should not be used
  --> tests/ui/global_variables.rs:11:9
   |
LL |         static GLOBAL_VARIABLE_1: Mutex<u32> = Mutex::new(3);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
LL |     define_global_variable_with_macro!();
   |     ------------------------------------ in this macro invocation
   |
   = note: this error originates in the macro `define_global_variable_with_macro` (in Nightly builds, run with -Z macro-backtrace for more info)

error: global variable should not be used
  --> tests/ui/global_variables.rs:20:5
   |
LL |     static GLOBAL_VARIABLE_2: AtomicU32 = AtomicU32::new(2);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider passing this as function arguments or using a `thread_local`

error: global variable should not be used
  --> tests/ui/global_variables.rs:23:5
   |
LL |     static GLOBAL_VARIABLE_3: Mutex<u32> = Mutex::new(3);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider passing this as function arguments or using a `thread_local`

error: global variable should not be used
  --> tests/ui/global_variables.rs:26:5
   |
LL |     static GLOBAL_VARIABLE_4: Option<AtomicU32> = Some(AtomicU32::new(0));
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider passing this as function arguments or using a `thread_local`

error: global variable should not be used
  --> tests/ui/global_variables.rs:47:13
   |
LL |             static GLOBAL_VARIABLE_IN_THREAD_LOCAL: AtomicU32 = AtomicU32::new(0);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider passing this as function arguments or using a `thread_local`

error: aborting due to 6 previous errors

